### ISSUE TITLE
setState inside setState keeps an unprocessed call in queue

### ISSUE THREAD

**Issue by unused** [facebook/react#8565]
**Do you want to request a *feature* or report a *bug*?**
*bug*

**What is the current behavior?**
A nested setState call is queued but only executed in next batch processing. Additionally, if any callback (2nd param) is set, this is executed immediately. 
Demo: https://jsfiddle.net/69z2wepo/65038/ - see console debug output

**What is the expected behavior?**
Everything in the queue should be processed, also if attached during execution.

Tested with current 15.4

**Comment by gaearon** [comment]
I don’t think calling `this.setState` from inside the updater argument was ever meant to be supported. What are you trying to accomplish? You need to *return* the next state from it, not to call `setState()` again. Ideally we should just throw there..

**Comment by unused** [comment]
TL;DR imho react does a great job offering simple, short, but complete documentation and useful warnings at (dev) runtime - it would be great to add a warning and a note in the documentation about this behaviour. Maybe it would be easy to fix it or on the other hand just prevent the current situation and ignore the call.

Thank you @gaearon for the quick reply. I totally agree that in most cases this won't be necessary. I'm using the media capture and media stream api a lot, it sometimes requires multiple actions taken to receive some result. While developing a new feature my colleagues recognized something was not  processed and I figured out it is because of a nested setState. Regarding to the [documentation](https://facebook.github.io/react/docs/state-and-lifecycle.html#using-state-correctly) I have to use a function as first argument to ensure I have the (at the time) state and props at hand and I cannot guarantee there is no other function triggered then. If it is not meant to be used inside the updater, I am sure to find a proper solution then.

**Comment by gaearon** [comment]
>it would be great to add a warning and a note in the documentation about this behaviour

I agree, it’s on our list.

>it sometimes requires multiple actions taken to receive some result.

I don't quite understand why you need to call `setState` from inside the updater.

When you're inside an updater and want to change the state you just need to return the next state.
Can you show me an example where you think `setState` would be necessary?

**Comment by unused** [comment]
Thank you a lot, if it should not be used this way, I'll take care to avoid it and looking forward to a friendly reminder :smile:. If some implementation requires the nested setState, I'll come up with an example.

**Comment by anvish** [comment]
I think React should treat nested setState calls as if they were sequential setState calls. 
This can be useful in an atomic state update. Simplified example:
```js
import React, { Component } from 'react';

class App extends Component {
  state = { count: 0 }
  click = () => {
    this.setState(state => {
      const newCount = state.count + 1
      this.load(newCount)
      return { count: newCount }
    })
  }
  load = (count) => {
    this.setState({ initialized: true, loading: true })
    fetch('something' + count).then(() => {
      //this.setState({ loading: false })
    })
  }
  render() {
    return (
      <div>
        <pre>{JSON.stringify(this.state)}</pre>
        <button type="button" onClick={this.click}>Go</button>
      </div>
    );
  }
}

export default App;
```
Second click or uncommenting the commented line leads to updating state.loading.

**Comment by bitstrider** [comment]
@anvish couldn't you achieve the same result by calling `setState` within the optional callback argument?
```javascript
import React, { Component } from 'react';

class App extends Component {
  state = { count: 0 }
  click = () => {
    this.setState(state => {
      const newCount = state.count + 1
      return { count: newCount, initialized: true, loading: true }
    }, () => {
      this.load(this.state.count)
    })
  }
  load = (count) => {
    fetch('something' + count).then(() => {
      this.setState({ loading: false })
    })
  }
...
```
